providers/providers.go

```diff
@@ -281,21 +282,23 @@ func (pm *ProviderManager) run() {
 					log.Error("error loading known provset: ", err)
 					continue
 				}
-				var filtered []peer.ID
 				for p, t := range provs.set {
-					if time.Now().Sub(t) > ProvideValidity {
+					if now.Sub(t) > ProvideValidity {
 						delete(provs.set, p)
-					} else {
-						filtered = append(filtered, p)
 					}
 				}
-
-				provs.providers = filtered
-				if len(filtered) == 0 {
+				// have we run out of providers?
+				if len(provs.set) == 0 {
+					provs.providers = nil
 					err := pm.deleteProvSet(k)
 					if err != nil {
 						log.Error("error deleting provider set: ", err)
 					}
+				} else if len(provs.providers) != len(provs.set) {
```

Comment:
What happens if the length is not different? I don't think you re-assign in this case.

Reply:
That means that we haven't removed any providers (I'm being lazy, I could keep an additional "modified" variable). I'll comment this.